### PR TITLE
#24778 add MergeLatest

### LINKED ISSUE #24778: MergeLatest List[Source[T]] => Source[List[T]] #1959
Moved from https://github.com/akka/akka-http/issues/1959 opened by @fmasion

Add MergeLast 

transform List[Source[T]] => Source[List[T]] 
The same way Future.sequence does with Future but while keeping the streaming nature of sources

A picture worth 1000 words :
https://blogs.endjin.com/wp-content/uploads/2014/05/combine-latest.png

here is the discussion on [gitter](https://gitter.im/akka/akka?at=5ab2ffb1458cbde55760b573)

here is a very naïve implementation [exemple gist](https://gist.github.com/fmasion/24707eaeb728ddaec8522989ed7d9b4d) (with no generic Traversable, canBuildfrom....)

Hope you find it useful

### PR DESCRIPTION
**Pull request by filipmatusak**
Resolves #24778

### CODE REVIEW

**Comment by hepin1989** [comment]
I have to say, I personally more prefer putting new `*GraphStage` implement to a dedicated file, not to make the `Graph.scala` file grow.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 1]
```diff
@@ -485,6 +486,75 @@ final class Interleave[T](val inputPorts: Int, val segmentSize: Int, val eagerCl
...
+object MergeLatest {
+  /**
+   * Create a new `MergeLatest` with the specified number of input ports.
+   *
+   * @param inputPorts number of input ports
+   * @param eagerComplete if true, the merge latest will complete as soon as one of its inputs completes.
+   */
+  def apply[T: ClassTag](inputPorts: Int, eagerComplete: Boolean = false): MergeLatest[T] = new MergeLatest[T](inputPorts, eagerComplete)
+
+}
```
[the remark this answers is not in the document]
Comment (filipmatusak) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: Do I need to transform GraphStage[UniformFanInShape[T, List[T]]] to GraphStage[UniformFanInShape[T, java.util.List[T]]] ?

[the remark this answers is not in the document]
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: Yes. Good catch, that should also be done, you can compose with `mapMaterializedValue` to transform the Scala list to a Java list

File: akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * MergeLatest joins elements from N input streams into stream of lists of size N.
+ * i-th element in list is the latest emitted element from i-th input stream.
+ * MergeLatest emits list for each element emitted from some input stream,
+ * but only after each stream emitted at least one element
+ *
+ * '''Emits when''' element is available from some input and each input emits at least one element from stream start
+ *
+ * '''Backpressures when''' downstream backpressures
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala]: Is that really true? I mean, the purpose is to have the _latest_ merged, which means that we do not want to backpressure but drop the elements while one has not pulled; otherwise such merge would not be any different from a normal Merge

File: akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * '''Completes when''' all upstreams complete (eagerClose=false) or one upstream completes (eagerClose=true)
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala]: good

File: akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ */
+object MergeLatest {
+  /**
+   * Create a new `MergeLatest` with the specified number of input ports.
+   *
+   * @param inputPorts number of input ports
+   * @param eagerComplete if true, the merge latest will complete as soon as one of its inputs completes.
+   */
+  def apply[T: ClassTag](inputPorts: Int, eagerComplete: Boolean = false): GraphStage[UniformFanInShape[T, java.util.List[T]]] =
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala]: do we need the classtag?

File: akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+package akka.stream.scaladsl
+
+import akka.stream.{ Attributes, Inlet, Outlet, UniformFanInShape }
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler }
+
+import scala.collection.{ immutable, mutable }
+import scala.reflect.ClassTag
+import scala.language.higherKinds
+
+object MergeLatest {
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala]: missing docs?

Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala]: Users will see docs of the companion in IDEs so having docs copied here or only here makes sense

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergeLatestSpec.scala [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+      val up2 = TestSource.probe[Int]
+      val up3 = TestSource.probe[Int]
+      val probe = TestSubscriber.manualProbe[List[Int]]()
+
+      val (in1, in2, in3) = RunnableGraph.fromGraph(GraphDSL.create(up1, up2, up3)((_, _, _)) { implicit b ⇒ (s1, s2, s3) ⇒
+        val m = b.add(MergeLatest[Int](3))
+
+        s1 ~> m.in(0)
+        s2 ~> m.in(1)
+        s3 ~> m.in(2)
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergeLatestSpec.scala]: A simple `~> m` should work AFAIK?

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergeLatestSpec.scala [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        m.out ~> Sink.fromSubscriber(probe)
+        ClosedShape
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergeLatestSpec.scala]: `SourceShape(out)` and then `runWith(TestSink.probe)` maybe?

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergeLatestSpec.scala [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+      val subscription = probe.expectSubscription()
+
+      subscription.request(1)
+      probe.expectNoMessage(10.millis)
+      in1.sendNext(1)
+      probe.expectNoMessage(10.millis)
+      in2.sendNext(2)
+      probe.expectNoMessage(10.millis)
+      in3.sendNext(3)
+      probe.expectNext(List(1, 2, 3))
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergeLatestSpec.scala]: This is not any different semantics than a normal merge though so far :)
Test should be:

```
in1.sendNext(1)
in2.sendNext(2)
in2.sendNext(22)
in3.sendNext(3)
in3.sendNext(33)
in3.sendNext(333)
```

Which should then when pulled end up with `List(1, 22, 333)` those are the "Latest" values

Comment (ktoso) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergeLatestSpec.scala]: Nevermind, I just saw you did such test in the next testcase. Sorry for the noise

Comment (ktoso) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergeLatestSpec.scala]: Reminding the image as well:



green 1 is sent
yellow 2 is sent
result 1 2
yellow 22 is sent
result 1 22
green 11 is sent
result 11 22

Comment (ktoso) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergeLatestSpec.scala]: Nevermind, I see you added a test for this in the below testcase :) Sorry for the noise

File: akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with OutHandler {
+    private val activeStreams: mutable.HashSet[Int] = new mutable.HashSet[Int]()
+    private var runningUpstreams: Int = inputPorts
+    private def upstreamsClosed: Boolean = runningUpstreams == 0
+    private def allMessagesReady: Boolean = activeStreams.size == inputPorts
+    private val messages: Array[T] = new Array[T](inputPorts)
+
+    override def preStart(): Unit = in.foreach(tryPull)
+
+    in.zipWithIndex.foreach {
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala]: should be in the preStart?

File: akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+ */
+final class MergeLatest[T: ClassTag, M](val inputPorts: Int, val eagerClose: Boolean)(buildElem: Array[T] ⇒ M) extends GraphStage[UniformFanInShape[T, M]] {
+  require(inputPorts >= 1, "input ports must be >= 1")
+
+  val in: immutable.IndexedSeq[Inlet[T]] = Vector.tabulate(inputPorts)(i ⇒ Inlet[T]("MergeLatest.in" + i))
+  val out: Outlet[M] = Outlet[M]("MergeLatest.out")
+  override val shape: UniformFanInShape[T, M] = UniformFanInShape(out, in: _*)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with OutHandler {
+    private val activeStreams: mutable.HashSet[Int] = new mutable.HashSet[Int]()
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala]: maybe use the java util one? AFAIR slightly lighter overhead

File: akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+            if (!eagerClose) {
+              runningUpstreams -= 1
+              if (upstreamsClosed) completeStage()
+            } else completeStage()
+          }
+        })
+    }
+
+    def onPull(): Unit =
+      for (index ← 0 until inputPorts)
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala]: pull in a while loop? allocating the range on every pull costs a bit allocation / performance wise

**Review by ktoso — COMMENTED** [review]
Looking good overall :) Thanks

Needs to be added to stages-overview.md doc as well so people can find it :)

**Review by ktoso — APPROVED** [review]

File: akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala (hunk 4 above)
Comment (filipmatusak) [reply akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala]: I need it for 'new Array[T]'

File: akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala (hunk 5 above)
Comment (filipmatusak) [reply akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala]: ok

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergeLatestSpec.scala (hunk 6 above)
Comment (filipmatusak) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergeLatestSpec.scala]: Yes, it works

**Comment by ktoso** [comment]
OK TO TEST

File: akka-docs/src/main/paradox/stream/stages-overview.md [hunk 12]
```diff
@@ -1766,6 +1766,19 @@ sources has elements ready the relative priorities for those sources are used to
...
 
+### mergeLatest
+
+MergeLatest joins elements from N input streams into stream of lists of size N.
+i-th element in list is the latest emitted element from i-th input stream.
+MergeLatest emits list for each element emitted from some input stream,
+but only after each stream emitted at least one element
+
+**emits** when element is available from some input and each input emits at least one element from stream start
+
```
Comment (ktoso) [inline akka-docs/src/main/paradox/stream/stages-overview.md]: should we mention explicitly that does not backpressure the upstreams?

File: akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala (hunk 4 above)
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala]: Got it, good 👍

**Review by ktoso — APPROVED** [review]

File: akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+              runningUpstreams -= 1
+              if (upstreamsClosed) completeStage()
+            } else completeStage()
+          }
+        })
+    }
+
+    def onPull(): Unit = {
+      var i = 0
+      while( i < inputPorts){
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala]: weird formatting, this will fail the build, please compile before pushing as it should reformat the code

File: akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala]: The Java API factory should be called `.create` and we need to explicitly accept the `Class[T]` as a parameter here, the Java compiler does not know of `ClassTag`s.

File: akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala [hunk 15]
```diff
@@ -31,8 +31,10 @@ object MergeLatest {
    * @param inputPorts number of input ports
    * @param eagerComplete if true, the merge latest will complete as soon as one of its inputs completes.
    */
-  def apply[T: ClassTag](inputPorts: Int, eagerComplete: Boolean = false): GraphStage[UniformFanInShape[T, java.util.List[T]]] =
+  def create[T](inputPorts: Int, eagerComplete: Boolean = false, clazz: Class[T]): GraphStage[UniformFanInShape[T, java.util.List[T]]] = {
+    implicit val classTag: ClassTag[T] = ClassTag(clazz)
     new MergeLatest[T, java.util.List[T]](inputPorts, eagerComplete)(x ⇒ x.toList.asJava)
+  }
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala]: Thinking about this some more, there must be a way to not require that class tag. The elements will anyways be boxed, so we could for example make the array contain `Any` in the stage and cast on the way out.

Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala]: Yeah one can make an AnyRef Array and cast the elements AFAIR?

File: akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ */
+object MergeLatest {
+  /**
+   * Create a new `MergeLatest` with the specified number of input ports.
+   *
+   * @param inputPorts number of input ports
+   * @param eagerComplete if true, the merge latest will complete as soon as one of its inputs completes.
+   */
+  def create[T](inputPorts: Int, eagerComplete: Boolean = false, clazz: Class[T]): GraphStage[UniformFanInShape[T, java.util.List[T]]] = {
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala]: I think this should be defined as separate overloaded `create` methods instead, similar to the `Merge`.

File: akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala (hunk 9 above)
Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/scaladsl/MergeLatest.scala]: I don't think that is necessary, handlers are normally set from constructor

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala [hunk 17]
```diff
@@ -25,6 +26,7 @@ import scala.concurrent.duration.FiniteDuration
 import scala.concurrent.{ Future, Promise }
 import akka.stream.stage.GraphStageWithMaterializedValue
 
+import scala.collection.generic.CanBuildFrom
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: unnecessary changes, probably not used imports

**Review by patriknw — CHANGES_REQUESTED** [review]
looking good, but javadsl should be adjusted

File: akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala [hunk 18]
```diff
@@ -31,10 +29,14 @@ object MergeLatest {
    * @param inputPorts number of input ports
    * @param eagerComplete if true, the merge latest will complete as soon as one of its inputs completes.
    */
-  def create[T](inputPorts: Int, eagerComplete: Boolean = false, clazz: Class[T]): GraphStage[UniformFanInShape[T, java.util.List[T]]] = {
-    implicit val classTag: ClassTag[T] = ClassTag(clazz)
-    new MergeLatest[T, java.util.List[T]](inputPorts, eagerComplete)(x ⇒ x.toList.asJava)
-  }
+  def apply[T](inputPorts: Int, eagerComplete: Boolean): GraphStage[UniformFanInShape[T, java.util.List[T]]] =
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/javadsl/MergeLatest.scala]: `create` instead of `apply` for these two factories

**Review by johanandren — COMMENTED** [review]
Good, simpler API without the class tag requirement

**Review by johanandren — APPROVED** [review]
LGTM, thank you!

**Review by johanandren — APPROVED** [review]
